activesupport/lib/active_support/callbacks.rb

```diff
@@ -86,21 +88,58 @@ module Callbacks
     #   run_callbacks :save do
     #     save
     #   end
+    #
+    #--
+    #
+    # As this method is used in many places, and often wraps large portions of
+    # user code, it has an additional design goal of minimizing its impact on
+    # the visible call stack. An exception from inside a :before or :after
+    # callback can be as noisy as it likes -- but when control has passed
+    # smoothly through and into the supplied block, we want as little evidence
+    # as possible that we were here.
     def run_callbacks(kind, &block)
-      send "_run_#{kind}_callbacks", &block
-    end
-
-    private
+      callbacks = __callbacks[kind.to_sym]
+
+      if callbacks.empty?
+        yield if block_given?
+      else
+        block ||= -> { true }
+        env = Filters::Environment.new(self, false, nil)
+        next_sequence = callbacks.compile
+
+        invoke_sequence = Proc.new do
+          skipped = nil
+          while true
+            current, next_sequence = next_sequence, next_sequence.nested
+            current.invoke_before(env)
+            if current.final?
+              env.value = !env.halted && block.call
+            elsif current.skip?(env)
+              (skipped ||= []) << current
+              next
+            else
+              expanded = current.expand_call_template(env, invoke_sequence)
+              expanded.shift.send(*expanded, &expanded.shift)
+            end
+            current.invoke_after(env)
+            skipped.pop.invoke_after(env) while skipped && skipped.first
+            break env.value
+          end
+        end
 
-      def __run_callbacks__(callbacks, &block)
-        if callbacks.empty?
-          yield if block_given?
+        # Common case: no 'around' callbacks defined
+        if next_sequence.final?
+          next_sequence.invoke_before(env)
+          env.value = !env.halted && block.call
+          next_sequence.invoke_after(env)
+          env.value
         else
-          runner = callbacks.compile
-          e = Filters::Environment.new(self, false, nil, block)
-          runner.call(e).value
+          invoke_sequence.call
```

Comment:
In-lining `invoke_sequence` here would eliminate the remaining double entry when an `around` is defined:

```
    lib/active_support/callbacks.rb:122:in `block in run_callbacks'
    lib/active_support/callbacks.rb:137:in `run_callbacks'
```

But that doesn't seem like a sensible trade.

(I actually like the symmetry with it there, anyway.)